lms/services/assignment.py

```diff
@@ -10,54 +12,199 @@ def __init__(self, db):
+            # ext_lti_assignment_id launch params.
+            assignments = self.get_for_canvas_launch(
+                tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+            )
+            if len(assignments) > 1:
+                raise MultipleResultsFound(
+                    "Multiple assignments found. Should merge_canvas_assignments have been called"
+                )
+            return assignments[0]
+
+        if not resource_link_id:
+            # When creating or editing an assignment Canvas launches us with an
+            # ext_lti_assignment_id but no resource_link_id.
+            return self._get_for_canvas_assignment_config(
+                tool_consumer_instance_guid, ext_lti_assignment_id
+            )
+
+        # Non-Canvas assignment launches always have a resource_link_id and never
+        # have an ext_lti_assignment_id.
+        return self._get_by_resource_link_id(
+            tool_consumer_instance_guid, resource_link_id
+        )
+
+    @lru_cache(maxsize=128)
+    def get_for_canvas_launch(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id,
+        ext_lti_assignment_id,
+    ):
+        """Get a canvas assignment by both resource_link_id and ext_lti_assignment_id."""
+        assert resource_link_id and ext_lti_assignment_id
+
+        assignments = (
+            self._db.query(Assignment)
+            .filter(
+                Assignment.tool_consumer_instance_guid == tool_consumer_instance_guid,
+                (
+                    (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                    | (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id.is_(None))
+                    )
+                    | (
+                        (Assignment.resource_link_id.is_(None))
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                ),
+            )
+            .order_by(Assignment.resource_link_id.asc())
+            .all()
+        )
+
+        if not assignments:
+            raise NoResultFound()
+
+        return assignments
+
+    def exists(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ) -> bool:
+        try:
+            self.get(
+                tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
+            )
+        except MultipleResultsFound:
+            # Merge needed but it exists
+            return True
+        except (NoResultFound, ValueError):
+            return False
+        else:
+            return True
+
+    def set_document_url(  # pylint:disable=too-many-arguments
+        self,
+        document_url,
```

Comment:
Args order changed. I think `document_url` makes sense as first or last argument but as I can't be mixed with other ones with default values and I don't think `document_url=None` makes sense I put it on front.

There's some changes on the launch views and test assertions because of this.